kratos/mpi/utilities/amgcl_distributed_csr_conversion_utilities.h

```diff
@@ -0,0 +1,167 @@
+		pAconverted->pGetColNumbering() = Kratos::make_unique<DistributedNumbering<IndexType>>(kratos_comm,rA.local()->ncols);
+
+		//here we fill the global to local mapping
+		const auto& comm_pattern = rA.cpat();
+		for(TIndexType i=0; i<rA.remote()->nnz; ++i){  //TODO: i suspect there is a smarter way to do this
+			TIndexType id = rA.remote()->col[i]; 
+			TIndexType local_id = comm_pattern.local_index(id);
+			pAconverted->GetOffDiagonalLocalIds()[id] = local_id; 
+			rA.remote()->col[i] = local_id; //note that here we overwrite the amgcl data!
+		}
+
+		//here we fill the local to global mapping (the inverse of the previous one)
+		pAconverted->GetOffDiagonalGlobalIds().resize(pAconverted->GetOffDiagonalLocalIds().size());
+
+		for(auto item : pAconverted->GetOffDiagonalLocalIds()){
+			pAconverted->GetOffDiagonalGlobalIds()[item.second] = item.first; //item.second=local_id, item.first=global_id
+		}
+
+		//setting col size for both matrices
+		pAconverted->GetDiagonalBlock().SetColSize(rA.local()->ncols); 
+		pAconverted->GetOffDiagonalBlock().SetColSize(pAconverted->GetOffDiagonalGlobalIds().size()); 
+
+		//convert diagonal block
+		pAconverted->pGetDiagonalBlock() = std::move(AmgclCSRConversionUtilities::ConvertToCsrMatrix<TDataType,TIndexType>(*(rA.local().get())));
+
+		//convert off diagonal block. Note that we need to change the usage of the index2. We do it in place
+		pAconverted->pGetOffDiagonalBlock() = std::move(AmgclCSRConversionUtilities::ConvertToCsrMatrix<TDataType,TIndexType>(*(rA.remote().get())));
+
+		//fill the vector importer, so that the matrix can be used to do calculations
+		pAconverted->pGetVectorImporter() = Kratos::make_unique<DistributedVectorImporter<TDataType,IndexType>>(
+			kratos_comm,pAconverted->GetOffDiagonalGlobalIds(), 
+			pAconverted->GetColNumbering()
+			); 
+
+		//the following data are simply not available in Amgcl.
+		//warning: THE IMPORTANT IMPLICATION IS THAT THE MATRIX CANNOT BE USED FOR ASSEMBLY WITHOUT PROVIDING MORE INFORMATION
+		// mSendCachedIJ
+        // mRecvCachedIJ
+        // mOffDiagonalLocalIds
+        // mOffDiagonalGlobalIds
+        // mfem_assemble_colors
+
+		//set ownership
+        // if(rA.local()->own_data == false){ //if rA is not the owner, Aconverted cannot be
+        //     pAconverted->GetDiagonalBlock().SetIsOwnerOfData(false);
+        // }
+        // else{ //if rA is the owner, transfer ownership to the csr_matrix
+        //     rA.local()->own_data = false;
+        //     pAconverted->GetDiagonalBlock().SetIsOwnerOfData(true);
+        // }		
+
+        // if(rA.remote()->own_data == false){ //if rA is not the owner, Aconverted cannot be
+        //     pAconverted->GetOffDiagonalBlock().SetIsOwnerOfData(false);
+        // }
+        // else{ //if rA is the owner, transfer ownership to the csr_matrix
+        //     rA.local()->own_data = false;
+        //     pAconverted->GetOffDiagonalBlock().SetIsOwnerOfData(true);
+        // }		
```

Comment:
Is this intentionally kept?

Reply:
i can remove much of it, however i would prefer to merge and to have another small PR just taking it out